File: src/client/acontext-py/src/acontext/resources/sessions.py
"""Sessions endpoints."""

import json
from collections.abc import Mapping
from dataclasses import asdict
from typing import Any, BinaryIO, Literal

from .._utils import build_params
from ..client_types import RequesterProtocol
from ..messages import AcontextMessage
from ..types.session import (
    GetMessagesOutput,
    GetTasksOutput,
    LearningStatus,
    ListSessionsOutput,
    Message,
    Session,
)
from ..uploads import FileUpload, normalize_file_upload
from pydantic import BaseModel
from openai.types.chat import ChatCompletionMessageParam
from anthropic.types import MessageParam

UploadPayload = (
    FileUpload | tuple[str, BinaryIO | bytes] | tuple[str, BinaryIO | bytes, str | None]
)
MessageBlob = AcontextMessage | ChatCompletionMessageParam | MessageParam


class SessionsAPI:
    def __init__(self, requester: RequesterProtocol) -> None:
        self._requester = requester

    def list(
        self,
        *,
        space_id: str | None = None,
        not_connected: bool | None = None,
        limit: int | None = None,
        cursor: str | None = None,
        time_desc: bool | None = None,
    ) -> ListSessionsOutput:
        """List all sessions in the project.

        Args:
            space_id: Filter sessions by space ID. Defaults to None.
            not_connected: Filter sessions that are not connected to a space. Defaults to None.
            limit: Maximum number of sessions to return. Defaults to None.
            cursor: Cursor for pagination. Defaults to None.
            time_desc: Order by created_at descending if True, ascending if False. Defaults to None.

        Returns:
            ListSessionsOutput containing the list of sessions and pagination information.
        """
        params: dict[str, Any] = {}
        if space_id:
            params["space_id"] = space_id
        params.update(
            build_params(
                not_connected=not_connected,
                limit=limit,
                cursor=cursor,
                time_desc=time_desc,
            )
        )
        data = self._requester.request("GET", "/session", params=params or None)
        return ListSessionsOutput.model_validate(data)

    def create(
        self,
        *,
        space_id: str | None = None,
        configs: Mapping[str, Any] | None = None,
    ) -> Session:
        """Create a new session.

        Args:
            space_id: Optional space ID to associate with the session. Defaults to None.
            configs: Optional session configuration dictionary. Defaults to None.

        Returns:
            The created Session object.
        """
        payload: dict[str, Any] = {}
        if space_id:
            payload["space_id"] = space_id
        if configs is not None:
            payload["configs"] = configs
        data = self._requester.request("POST", "/session", json_data=payload)
        return Session.model_validate(data)

    def delete(self, session_id: str) -> None:
        """Delete a session by its ID.

        Args:
            session_id: The UUID of the session to delete.
        """
        self._requester.request("DELETE", f"/session/{session_id}")

    def update_configs(
        self,
        session_id: str,
        *,
        configs: Mapping[str, Any],
    ) -> None:
        """Update session configurations.

        Args:
            session_id: The UUID of the session.
            configs: Session configuration dictionary.
        """
        payload = {"configs": configs}
        self._requester.request(
            "PUT", f"/session/{session_id}/configs", json_data=payload
        )

    def get_configs(self, session_id: str) -> Session:
        """Get session configurations.

        Args:
            session_id: The UUID of the session.

        Returns:
            Session object containing the configurations.
        """
        data = self._requester.request("GET", f"/session/{session_id}/configs")
        return Session.model_validate(data)

    def connect_to_space(self, session_id: str, *, space_id: str) -> None:
        """Connect a session to a space.

        Args:
            session_id: The UUID of the session.
            space_id: The UUID of the space to connect to.
        """
        payload = {"space_id": space_id}
        self._requester.request(
            "POST", f"/session/{session_id}/connect_to_space", json_data=payload
        )

    def get_tasks(
        self,
        session_id: str,
        *,
        limit: int | None = None,
        cursor: str | None = None,
        time_desc: bool | None = None,
    ) -> GetTasksOutput:
        """Get tasks for a session.

        Args:
            session_id: The UUID of the session.
            limit: Maximum number of tasks to return. Defaults to None.
            cursor: Cursor for pagination. Defaults to None.
            time_desc: Order by created_at descending if True, ascending if False. Defaults to None.

        Returns:
            GetTasksOutput containing the list of tasks and pagination information.
        """
        params = build_params(limit=limit, cursor=cursor, time_desc=time_desc)
        data = self._requester.request(
            "GET",
            f"/session/{session_id}/task",
            params=params or None,
        )
        return GetTasksOutput.model_validate(data)

    def send_message(
        self,
        session_id: str,
        *,
        blob: MessageBlob,
        format: Literal["acontext", "openai", "anthropic"] = "openai",
        file_field: str | None = None,
        file: (
            FileUpload
            | tuple[str, BinaryIO | bytes]
            | tuple[str, BinaryIO | bytes, str]
            | None
        ) = None,
    ) -> Message:
        """Send a message to a session.

        Args:
            session_id: The UUID of the session.
            blob: The message blob in Acontext, OpenAI, or Anthropic format.
            format: The format of the message blob. Defaults to "openai".
            file_field: The field name for file upload. Only used when format is "acontext".
                Required if file is provided. Defaults to None.
            file: Optional file upload. Only used when format is "acontext". Defaults to None.

        Returns:
            The created Message object.

        Raises:
            ValueError: If format is invalid, file/file_field provided for non-acontext format,
                or file is provided without file_field for acontext format.
        """
        if format not in {"acontext", "openai", "anthropic"}:
            raise ValueError(
                "format must be one of {'acontext', 'openai', 'anthropic'}"
            )

        # File upload is only supported for acontext format
        if format != "acontext" and (file is not None or file_field is not None):
            raise ValueError(
                "file and file_field parameters are only supported when format is 'acontext'"
            )
        if isinstance(blob, BaseModel):
            blob = blob.model_dump()

        payload: dict[str, Any] = {
            "format": format,
        }
        if format == "acontext":
            if isinstance(blob, Mapping):
                payload["blob"] = blob
            elif isinstance(blob, AcontextMessage):
                payload["blob"] = asdict(blob)
            else:
                raise ValueError(
                    f"Invalid blob type: {type(blob)} when format is 'acontext'. Expected Mapping or AcontextMessage"
                )

            # Handle file upload for acontext format
            file_payload: dict[str, tuple[str, BinaryIO, str]] | None = None
            if file is not None:
                if file_field is None:
                    raise ValueError("file_field is required when file is provided")
                # only support upload one file now
                upload = normalize_file_upload(file)
                file_payload = {file_field: upload.as_httpx()}

            if file_payload:
                form_data = {"payload": json.dumps(payload)}
                data = self._requester.request(
                    "POST",
                    f"/session/{session_id}/messages",
                    data=form_data,
                    files=file_payload,
                )
            else:
                data = self._requester.request(
                    "POST",
                    f"/session/{session_id}/messages",
                    json_data=payload,
                )
        else:
            payload["blob"] = blob  # type: ignore
            data = self._requester.request(
                "POST",
                f"/session/{session_id}/messages",
                json_data=payload,
            )
        return Message.model_validate(data)

    def get_messages(
        self,
        session_id: str,
        *,
        limit: int | None = None,
        cursor: str | None = None,
        with_asset_public_url: bool | None = None,
        format: Literal["acontext", "openai", "anthropic"] = "openai",
        time_desc: bool | None = None,
    ) -> GetMessagesOutput:
        """Get messages for a session.

        Args:
            session_id: The UUID of the session.
            limit: Maximum number of messages to return. Defaults to None.
            cursor: Cursor for pagination. Defaults to None.
            with_asset_public_url: Whether to include presigned URLs for assets. Defaults to None.
            format: The format of the messages. Defaults to "acontext".
            time_desc: Order by created_at descending if True, ascending if False. Defaults to None.

        Returns:
            GetMessagesOutput containing the list of messages and pagination information.
        """
        params: dict[str, Any] = {}
        if format is not None:
            params["format"] = format
        params.update(
            build_params(
                limit=limit,
                cursor=cursor,
                with_asset_public_url=with_asset_public_url,
                time_desc=time_desc,
            )
        )
        data = self._requester.request(
            "GET", f"/session/{session_id}/messages", params=params or None
        )
        return GetMessagesOutput.model_validate(data)

    def flush(self, session_id: str) -> dict[str, Any]:
        """Flush the session buffer for a given session.

        Args:
            session_id: The UUID of the session.

        Returns:
            Dictionary containing status and errmsg fields.
        """
        data = self._requester.request("POST", f"/session/{session_id}/flush")
        return data  # type: ignore

    def get_learning_status(self, session_id: str) -> LearningStatus:
        """Get learning status for a session.

        Returns the count of space digested tasks and not space digested tasks.
        If the session is not connected to a space, returns 0 and 0.

        Args:
            session_id: The UUID of the session.

        Returns:
            LearningStatus object containing space_digested_count and not_space_digested_count.
        """
        data = self._requester.request(
            "GET", f"/session/{session_id}/get_learning_status"
        )
        return LearningStatus.model_validate(data)

File: src/client/acontext-ts/src/resources/sessions.ts
/**
 * Sessions endpoints.
 */

import { RequesterProtocol } from '../client-types';
import { AcontextMessage, AcontextMessageInput } from '../messages';
import { FileUpload } from '../uploads';
import { buildParams } from '../utils';
import {
  GetMessagesOutput,
  GetMessagesOutputSchema,
  GetTasksOutput,
  GetTasksOutputSchema,
  ListSessionsOutput,
  ListSessionsOutputSchema,
  Message,
  MessageSchema,
  Session,
  SessionSchema,
} from '../types';

export type MessageBlob = AcontextMessage | Record<string, unknown>;

export class SessionsAPI {
  constructor(private requester: RequesterProtocol) {}

  async list(options?: {
    spaceId?: string | null;
    notConnected?: boolean | null;
    limit?: number | null;
    cursor?: string | null;
    timeDesc?: boolean | null;
  }): Promise<ListSessionsOutput> {
    const params: Record<string, string | number> = {};
    if (options?.spaceId) {
      params.space_id = options.spaceId;
    }
    Object.assign(
      params,
      buildParams({
        not_connected: options?.notConnected ?? null,
        limit: options?.limit ?? null,
        cursor: options?.cursor ?? null,
        time_desc: options?.timeDesc ?? null,
      })
    );
    const data = await this.requester.request('GET', '/session', {
      params: Object.keys(params).length > 0 ? params : undefined,
    });
    return ListSessionsOutputSchema.parse(data);
  }

  async create(options?: {
    spaceId?: string | null;
    configs?: Record<string, unknown>;
  }): Promise<Session> {
    const payload: Record<string, unknown> = {};
    if (options?.spaceId) {
      payload.space_id = options.spaceId;
    }
    if (options?.configs !== undefined) {
      payload.configs = options.configs;
    }
    const data = await this.requester.request('POST', '/session', {
      jsonData: Object.keys(payload).length > 0 ? payload : undefined,
    });
    return SessionSchema.parse(data);
  }

  async delete(sessionId: string): Promise<void> {
    await this.requester.request('DELETE', `/session/${sessionId}`);
  }

  async updateConfigs(
    sessionId: string,
    options: {
      configs: Record<string, unknown>;
    }
  ): Promise<void> {
    const payload = { configs: options.configs };
    await this.requester.request('PUT', `/session/${sessionId}/configs`, {
      jsonData: payload,
    });
  }

  async getConfigs(sessionId: string): Promise<Session> {
    const data = await this.requester.request('GET', `/session/${sessionId}/configs`);
    return SessionSchema.parse(data);
  }

  async connectToSpace(
    sessionId: string,
    options: {
      spaceId: string;
    }
  ): Promise<void> {
    const payload = { space_id: options.spaceId };
    await this.requester.request('POST', `/session/${sessionId}/connect_to_space`, {
      jsonData: payload,
    });
  }

  async getTasks(
    sessionId: string,
    options?: {
      limit?: number | null;
      cursor?: string | null;
      timeDesc?: boolean | null;
    }
  ): Promise<GetTasksOutput> {
    const params = buildParams({
      limit: options?.limit ?? null,
      cursor: options?.cursor ?? null,
      time_desc: options?.timeDesc ?? null,
    });
    const data = await this.requester.request('GET', `/session/${sessionId}/task`, {
      params: Object.keys(params).length > 0 ? params : undefined,
    });
    return GetTasksOutputSchema.parse(data);
  }

  async sendMessage(
    sessionId: string,
    blob: MessageBlob,
    options?: {
      format?: 'acontext' | 'openai' | 'anthropic';
      fileField?: string | null;
      file?: FileUpload | null;
    }
  ): Promise<Message> {
    const format = options?.format ?? 'openai';
    if (!['acontext', 'openai', 'anthropic'].includes(format)) {
      throw new Error("format must be one of {'acontext', 'openai', 'anthropic'}");
    }

    if (options?.file && !options?.fileField) {
      throw new Error('fileField is required when file is provided');
    }

    const payload: Record<string, unknown> = {
      format,
    };

    if (format === 'acontext') {
      if (blob instanceof AcontextMessage) {
        payload.blob = blob.toJSON();
      } else {
        // Try to parse as AcontextMessageInput
        // MessageBlob can be Record<string, unknown>, which may not match AcontextMessageInput exactly
        const message = new AcontextMessage(blob as AcontextMessageInput);
        payload.blob = message.toJSON();
      }
    } else {
      payload.blob = blob;
    }

    if (options?.file && options?.fileField) {
      const formData: Record<string, string> = {
        payload: JSON.stringify(payload),
      };
      const files = {
        [options.fileField]: options.file.asFormData(),
      };
      const data = await this.requester.request('POST', `/session/${sessionId}/messages`, {
        data: formData,
        files,
      });
      return MessageSchema.parse(data);
    } else {
      const data = await this.requester.request('POST', `/session/${sessionId}/messages`, {
        jsonData: payload,
      });
      return MessageSchema.parse(data);
    }
  }

  async getMessages(
    sessionId: string,
    options?: {
      limit?: number | null;
      cursor?: string | null;
      withAssetPublicUrl?: boolean | null;
      format?: 'acontext' | 'openai' | 'anthropic';
      timeDesc?: boolean | null;
    }
  ): Promise<GetMessagesOutput> {
    const params: Record<string, string | number> = {};
    if (options?.format !== undefined) {
      params.format = options.format;
    }
    Object.assign(
      params,
      buildParams({
        limit: options?.limit ?? null,
        cursor: options?.cursor ?? null,
        with_asset_public_url: options?.withAssetPublicUrl ?? null,
        time_desc: options?.timeDesc ?? null,
      })
    );
    const data = await this.requester.request('GET', `/session/${sessionId}/messages`, {
      params: Object.keys(params).length > 0 ? params : undefined,
    });
    return GetMessagesOutputSchema.parse(data);
  }
}

